Tests/SourceKitLSPTests/TypeHierarchyTests.swift

```diff
@@ -0,0 +1,167 @@
+      Location(badUTF16: ws.testLoc(name))
+    }
+
+    func withoutData(_ item: TypeHierarchyItem) -> TypeHierarchyItem {
+      var item = item
+      item.data = nil
+      return item
+    }
+
+    /// Compares the given type hierarchies ignoring the implementation-specific
+    /// data field (which includes e.g. USRs that are difficult to test, especially
+    /// in the presence of extensions, and are not user-visible anyway).
+    func assertEqualIgnoringData(
+      _ actual: [TypeHierarchyItem],
+      _ expected: [TypeHierarchyItem],
+      file: StaticString = #file,
+      line: UInt = #line
+    ) {
+      XCTAssertEqual(
+        actual.map(withoutData),
+        expected.map(withoutData),
+        file: file,
+        line: line
+      )
+    }
+
+    func item(_ name: String, _ kind: SymbolKind, detail: String = "main", at locName: String) -> TypeHierarchyItem {
+      let location = loc(locName)
+      return TypeHierarchyItem(
+        name: name,
+        kind: kind,
+        tags: nil,
+        detail: detail,
+        uri: location.uri,
+        range: location.range,
+        selectionRange: location.range
+      )
+    }
+
+    // Test type hierarchy preparation
+
+    assertEqualIgnoringData(try typeHierarchy(at: testLoc("P")), [
+      item("P", .interface, at: "P"),
+    ])
+    assertEqualIgnoringData(try typeHierarchy(at: testLoc("A")), [
+      item("A", .class, at: "A"),
+    ])
+    assertEqualIgnoringData(try typeHierarchy(at: testLoc("S")), [
+      item("S", .struct, at: "S"),
+    ])
+    assertEqualIgnoringData(try typeHierarchy(at: testLoc("E")), [
+      item("E", .enum, at: "E"),
+    ])
+
+    // Test supertype hierarchy
+
+    assertEqualIgnoringData(try supertypes(at: testLoc("A")), [])
+    assertEqualIgnoringData(try supertypes(at: testLoc("B")), [
+      item("A", .class, at: "A"),
+      item("P", .interface, at: "P"),
+    ])
+    assertEqualIgnoringData(try supertypes(at: testLoc("C")), [
+      item("B", .class, at: "B"),
+    ])
+    assertEqualIgnoringData(try supertypes(at: testLoc("D")), [
+      item("A", .class, at: "A"),
+    ])
+    assertEqualIgnoringData(try supertypes(at: testLoc("S")), [
+      item("P", .interface, at: "P"),
+    ])
+    assertEqualIgnoringData(try supertypes(at: testLoc("E")), [
+      item("P", .interface, at: "P"),
+    ])
+
+    // Test subtype hierarchy (includes extensions)
+
+    assertEqualIgnoringData(try subtypes(at: testLoc("A")), [
+      item("B", .class, at: "B"),
+      item("D", .class, at: "D"),
+    ])
+    assertEqualIgnoringData(try subtypes(at: testLoc("B")), [
+      item("C", .class, at: "C"),
+    ])
+    assertEqualIgnoringData(try subtypes(at: testLoc("P")), [
+      item("B", .class, at: "B"),
+      item("S", .struct, at: "S"),
+      item("E", .enum, at: "E"),
+    ])
+    assertEqualIgnoringData(try subtypes(at: testLoc("S")), [
+      item("S: X", .null, detail: "Extension at a.swift:13", at: "extS:X"),
```

Comment:
I think an extension that adds multiple protocol conformances would also be interesting.

E.g. 
```swift
struct T {}
extension T: X, P {}
```

I would expect the extension to be displayed as `T: X, P`.

Also: I really like how you show the conformed types in the extension names. Makes it much easier to find the extension you are looking for IMO.